src/btree/bt_walk.c

```diff
@@ -304,10 +325,25 @@ __wt_tree_walk(WT_SESSION_IMPL *session,
 					break;
 			}
 
-			/* Optionally skip leaf pages. */
-			if (LF_ISSET(WT_READ_SKIP_LEAF) &&
-			    __wt_ref_is_leaf(session, ref))
-				break;
+			/*
+			 * Optionally skip leaf pages. If WT_READ_SKIP_LEAF is
+			 * set, we're passed in a count of leaf pages to skip.
+			 * If this page is disk-based, we can crack the cell
+			 * and figure out it's a leaf page without reading it.
+			 * Decrement the count of pages to zero, and then take
+			 * the next leaf page we can; be cautious around the
+			 * page decrement: if for some reason we end up not
+			 * taking this particular page, we can take the next
+			 * one, and, there are additional tests/decrements when
+			 * we're about to return a leaf page.
+			 */
+			if (LF_ISSET(WT_READ_SKIP_LEAF)) {
+				WT_ERR(__ref_is_leaf(session, ref, &isleaf));
+				if (isleaf && *walkcntp > 0) {
+					--*walkcntp;
```

Comment:
I'm not a fan of in/out parameters if we can avoid it.  A separate "skip_count" (input) parameter would be better.